File: apps/api/src/application/products/get-product.ts
import { NotFoundError } from "@/domain/shared/domain-error";

import type { ProductRepository } from "@/application/ports/product-repository";
import type { Product } from "@/domain/product/product";

export class GetProduct {
  constructor(private readonly products: ProductRepository) {}

  async execute(id: string): Promise<Product> {
    const product = await this.products.findById(id);
    if (!product) {
      throw new NotFoundError("product", id);
    }
    return product;
  }
}

File: apps/api/src/ecommerce_api/application/products/get_product.py
from ecommerce_api.application.ports.product_repository import ProductRepository
from ecommerce_api.domain.errors import NotFoundError
from ecommerce_api.domain.product import Product


class GetProduct:
    def __init__(self, products: ProductRepository) -> None:
        self._products = products

    async def execute(self, id: str) -> Product:
        product = await self._products.find_by_id(id)
        if product is None:
            raise NotFoundError('product', id)
        return product
